src/lib.rs

```diff
@@ -29,62 +40,68 @@ impl <const W: usize, const H: usize> Content<W, H> {
 /// This receives the `SmartLedsWriter` trait implementations along with a
 /// `MatrixType` that describes the size and the pixels mapping between the LED
 /// strip placement and the matrix's x y coordinates.
-pub struct SmartLedMatrix<T, M: MatrixType, const W: usize, const H: usize> {
+pub struct SmartLedMatrix<T, M: Transformation<W, H>, const W: usize, const H: usize> {
     writer: T,
     content: Content<W, H>,
-    matrix_type: M,
+    matrix_type: PhantomData<M>,
     brightness: u8,
 }
 
-impl<T, M: MatrixType, const W: usize, const H: usize> SmartLedMatrix<T, M, W, H> {
+impl<T, M: Transformation<W, H>, const W: usize, const H: usize> SmartLedMatrix<T, M, W, H> {
     pub fn set_brightness(&mut self, new_brightness: u8) {
         self.brightness = new_brightness;
     }
 }
 
-impl<T: SmartLedsWrite, M: MatrixType, const W: usize, const H: usize> OriginDimensions for SmartLedMatrix<T, M, W, H> {
+impl<T: SmartLedsWrite, M: Transformation<W, H>, const W: usize, const H: usize> OriginDimensions
+    for SmartLedMatrix<T, M, W, H>
+{
     fn size(&self) -> Size {
-        self.matrix_type.size()
+        Size::new(W as u32, H as u32)
     }
 }
 
-impl<T: SmartLedsWrite, M: MatrixType, const W: usize, const H: usize> SmartLedMatrix<T, M, W, H> 
-where <T as SmartLedsWrite>::Color: From<RGB8> {
-    pub fn new(writer: T, matrix_type: M) -> Self {
+impl<T: SmartLedsWrite, M: Transformation<W, H>, const W: usize, const H: usize>
+    SmartLedMatrix<T, M, W, H>
+where
+    <T as SmartLedsWrite>::Color: From<RGB8>,
+{
+    pub fn new(writer: T) -> Self {
         let content = Content::<W, H>([[RGB8::default(); W]; H]);
-        Self{writer: writer,
-            content: content,
-            matrix_type: matrix_type,
-            brightness: 255}
+        Self {
+            writer,
+            content,
+            matrix_type: PhantomData,
+            brightness: 255,
+        }
     }
     pub fn flush(&mut self) -> Result<(), DisplayError> {
         let iter = brightness(self.content.as_slice().iter().cloned(), self.brightness);
-        match self.writer.write(iter) {
-            Ok(()) => {
-                Ok(())
-            }
-            Err(_) => {
-                Err(DisplayError::BusWriteError)
-            }
-        }        
+        self.writer
+            .write(iter)
+            .map_err(|_| DisplayError::BusWriteError)
```

Comment:
`flush` could return `Result<(), T::Error>` instead of mapping it to a `DisplayError`. This also means that you can get rid of the `display-interface` dependency.